core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/CsdLocationFilter.java

```diff
@@ -1,33 +1,64 @@
 package com.sequenceiq.cloudbreak.service.upgrade.image;
 
+import java.util.Collections;
 import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.stereotype.Component;
+import org.springframework.util.CollectionUtils;
 
-import com.sequenceiq.cloudbreak.cloud.model.catalog.Image;
 import com.sequenceiq.cloudbreak.api.endpoint.v4.common.StackType;
+import com.sequenceiq.cloudbreak.cloud.model.catalog.Image;
 
 @Component
 public class CsdLocationFilter implements PackageLocationFilter {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(CsdLocationFilter.class);
 
     @Override
-    public boolean filterImage(Image image, Image currentImage, StackType stackType) {
-        if (StackType.WORKLOAD.equals(stackType)) {
+    public boolean filterImage(Image image, Image currentImage, ImageFilterParams imageFilterParams) {
+        if (StackType.WORKLOAD.equals(imageFilterParams.getStackType())) {
             if (image == null || image.getPreWarmCsd() == null || image.getPreWarmCsd().isEmpty()) {
                 LOGGER.debug("Image or CSD parcels are not present. Image: {}", image);
                 return false;
             } else {
-                List<String> csdList = image.getPreWarmCsd();
+                List<String> csdList = getPreWarmCsdList(image, imageFilterParams);
                 LOGGER.debug("Matching URLs: [{}]", csdList);
-                return csdList.stream().allMatch(csdUrl -> URL_PATTERN.matcher(csdUrl).find());
+                return !csdList.isEmpty() && csdList.stream().allMatch(csdUrl -> URL_PATTERN.matcher(csdUrl).find());
             }
         } else {
-            LOGGER.debug("Skip filtering because the stack type is {}", stackType);
+            LOGGER.debug("Skip filtering because the stack type is {}", imageFilterParams.getStackType());
             return false;
         }
     }
+
+    private List<String> getPreWarmCsdList(Image image, ImageFilterParams imageFilterParams) {
+        Set<String> activatedParcelNames = getParcelNames(imageFilterParams);
+        List<String> imagePreWarmCsd = image.getPreWarmCsd();
+        return CollectionUtils.isEmpty(activatedParcelNames) ? imagePreWarmCsd
+                : filterPreWarmCsdListByActivatedParcels(activatedParcelNames, imagePreWarmCsd);
+    }
+
+    private List<String> filterPreWarmCsdListByActivatedParcels(Set<String> activatedParcelNames, List<String> imagePreWarmCsd) {
```

Comment:
Renaming suggestion to accommodate previous renames: filterPreWarmCsdListByRelatedParcels